precise/iaf_psc_alpha_ps.cpp

```diff
@@ -0,0 +1,832 @@
+  def< double >( d, names::t_ref, t_ref_ );
+  def< long >( d, names::Interpol_Order, Interpol_ );
+}
+
+double
+nest::iaf_psc_alpha_ps::Parameters_::set( const DictionaryDatum& d )
+{
+  // if E_L_ is changed, we need to adjust all variables defined relative to
+  // E_L_
+  const double ELold = E_L_;
+  updateValue< double >( d, names::E_L, E_L_ );
+  const double delta_EL = E_L_ - ELold;
+
+  updateValue< double >( d, names::tau_m, tau_m_ );
+  updateValue< double >( d, names::tau_syn_ex, tau_syn_ex_ );
+  updateValue< double >( d, names::tau_syn_in, tau_syn_in_ );
+  updateValue< double >( d, names::C_m, c_m_ );
+  updateValue< double >( d, names::t_ref, t_ref_ );
+  updateValue< double >( d, names::I_e, I_e_ );
+
+  if ( updateValue< double >( d, names::V_th, U_th_ ) )
+  {
+    U_th_ -= E_L_;
+  }
+  else
+  {
+    U_th_ -= delta_EL;
+  }
+
+  if ( updateValue< double >( d, names::V_min, U_min_ ) )
+  {
+    U_min_ -= E_L_;
+  }
+  else
+  {
+    U_min_ -= delta_EL;
+  }
+
+  if ( updateValue< double >( d, names::V_reset, U_reset_ ) )
+  {
+    U_reset_ -= E_L_;
+  }
+  else
+  {
+    U_reset_ -= delta_EL;
+  }
+
+  long tmp;
+  if ( updateValue< long >( d, names::Interpol_Order, tmp ) )
+  {
+    if ( NO_INTERPOL <= tmp && tmp < END_INTERP_ORDER )
+    {
+      Interpol_ = static_cast< interpOrder >( tmp );
+    }
+    else
+    {
+      throw BadProperty(
+        "Invalid interpolation order. "
+        "Valid orders are 0, 1, 2, 3." );
+    }
+  }
+  if ( U_reset_ >= U_th_ )
+  {
+    throw BadProperty( "Reset potential must be smaller than threshold." );
+  }
+  if ( U_reset_ < U_min_ )
+  {
+    throw BadProperty(
+      "Reset potential must be greater equal minimum potential." );
+  }
+  if ( c_m_ <= 0 )
+  {
+    throw BadProperty( "Capacitance must be strictly positive." );
+  }
+
+  if ( Time( Time::ms( t_ref_ ) ).get_steps() < 1 )
+  {
+    throw BadProperty( "Refractory time must be at least one time step." );
+  }
+  if ( tau_m_ <= 0 || tau_syn_ex_ <= 0 || tau_syn_in_ <= 0 )
+  {
+    throw BadProperty( "All time constants must be strictly positive." );
+  }
+
+  return delta_EL;
+}
+
+void
+nest::iaf_psc_alpha_ps::State_::get( DictionaryDatum& d,
+  const Parameters_& p ) const
+{
+  def< double >( d, names::V_m, V_m_ + p.E_L_ ); // Membrane potential
+  def< double >( d, names::I_syn_ex, I_ex_ );    // Excitatory synaptic current
+  def< double >( d, names::I_syn_in, I_in_ );    // Inhibitory synaptic current
+  def< double >( d, names::dI_syn_in, dI_in_ );  // Derivative inhib. current
+  def< double >( d, names::dI_syn_ex, dI_ex_ );  // Derivative exc. current
+  def< bool >( d, names::is_refractory, is_refractory_ );
+}
+
+void
+nest::iaf_psc_alpha_ps::State_::set( const DictionaryDatum& d,
+  const Parameters_& p,
+  double delta_EL )
+{
+  if ( updateValue< double >( d, names::V_m, V_m_ ) )
+  {
+    V_m_ -= p.E_L_;
+  }
+  else
+  {
+    V_m_ -= delta_EL;
+  }
+
+  updateValue< double >( d, names::I_syn_ex, I_ex_ );
```

Comment:
The synaptic currents and their derivatives should not be settable. I don't think they are in any other models.

Reply:
They were in [canon](https://github.com/nest/nest-simulator/blob/master/precise/iaf_psc_alpha_canon.cpp#L206) but this is set state, not parameters, so it seems normal to me

Reply:
Your are right about `canon`, but I think that is rather a leftover from debugging. The derivatives are not documented in `canon`, and they cannot be set in the plain `iaf_psc_...` models. So I would argue against introducing them here.